### PR TITLE
gRPC handler testing

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: cartographer_grpc/framework/server.h [hunk 1]
```diff
@@ -104,9 +104,10 @@ class Server {
     return {execution_context_->lock(), execution_context_.get()};
   }
 
- private:
+ protected:
```
Comment (gaschler) [inline cartographer_grpc/framework/server.h]: You only need to expose AddService? Maybe leave the rest private.

File: cartographer_grpc/framework/testing/rpc_handler_wrapper.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace cartographer_grpc {
+namespace framework {
+namespace testing {
+
+template <class RpcHandlerType>
+class RpcHandlerWrapper : public RpcHandlerType {
+ public:
+  enum RpcHandlerEvent { ON_REQUEST, ON_READS_DONE, ON_FINISH };
+  using EventPublisher = std::function<void(RpcHandlerEvent)>;
```
Comment (gaschler) [inline cartographer_grpc/framework/testing/rpc_handler_wrapper.h]: naming: (Event)Callback?

File: cartographer_grpc/framework/testing/rpc_handler_test_server.h [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+const std::string kServerAddress = "localhost:50051";
+}  // namespace
+
+template <class RpcHandlerType>
+class RpcHandlerTestServer : public Server {
+ public:
+  RpcHandlerTestServer() : Server(Options{1, 1, kServerAddress}) {
+    // Register the handler under test.
+    this->AddService(kServiceName, {{kMethodName, GetRpcHandlerInfo()}});
+
```
Comment (gaschler) [inline cartographer_grpc/framework/testing/rpc_handler_test_server.h]: remove blank lines

File: cartographer_grpc/framework/testing/rpc_handler_test_server.h [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    // event in the handler under test.
+    InstantiateReadersAndWriters();
+  }
+
+  ~RpcHandlerTestServer() { this->Shutdown(); };
+
+  // Parses a request message from the passed string and issues the
+  // request against the handler, waits for the handler to complete
+  // processing before returning.
+  void SendMessage(const std::string &serialized_message) {
```
Comment (gaschler) [inline cartographer_grpc/framework/testing/rpc_handler_test_server.h]: You don't parse serialized_message.

Comment (gaschler) [inline cartographer_grpc/framework/testing/rpc_handler_test_server.h]: naming consistency: SendWrite

File: cartographer_grpc/framework/testing/rpc_handler_test_server.h [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  std::unique_ptr<ClientWriter> CreateClientWriter() {
+    return std::unique_ptr<ClientWriter>(ClientWriter::Create(
+        channel_.get(), rpc_method_, &context_, response_.get()));
+  }
+
+  RpcHandlerInfo GetRpcHandlerInfo() {
+    ::grpc::internal::RpcMethod::RpcType rpc_type =
+        RpcType<typename RpcHandlerType::IncomingType,
+                typename RpcHandlerType::OutgoingType>::value;
+    return RpcHandlerInfo{
```
Comment (gaschler) [inline cartographer_grpc/framework/testing/rpc_handler_test_server.h]: move the event publisher and handler instantiator here:
auto event_publisher = [this]....
auto handler_instantiator = [this]...

File: cartographer_grpc/framework/testing/rpc_handler_test_server.h [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        LOG(FATAL) << "Not implemented";
+        break;
+    }
+    WaitForHandlerCompletion(RpcHandlerWrapper::ON_FINISH);
+  }
+
+ private:
+  using ClientWriter = ::grpc::internal::ClientWriterFactory<
+      typename RpcHandlerType::RequestType>;
+  using HandlerInstantiator =
```
Comment (gaschler) [inline cartographer_grpc/framework/testing/rpc_handler_test_server.h]: remove this, use auto

File: cartographer_grpc/framework/server.h (hunk 1 above)
Comment (cschuet) [reply cartographer_grpc/framework/server.h]: Actually the destructor needs to be public and I need the constructor to be protected as well.

Comment (cschuet) [reply cartographer_grpc/framework/server.h]: Done.

File: cartographer_grpc/framework/testing/rpc_handler_test_server.h (hunk 3 above)
Comment (cschuet) [reply cartographer_grpc/framework/testing/rpc_handler_test_server.h]: Done.

File: cartographer_grpc/framework/testing/rpc_handler_test_server.h (hunk 4 above)
Comment (cschuet) [reply cartographer_grpc/framework/testing/rpc_handler_test_server.h]: Done.

Comment (cschuet) [reply cartographer_grpc/framework/testing/rpc_handler_test_server.h]: Thanks!

File: cartographer_grpc/framework/testing/rpc_handler_test_server.h (hunk 6 above)
Comment (cschuet) [reply cartographer_grpc/framework/testing/rpc_handler_test_server.h]: DOne.

File: cartographer_grpc/framework/testing/rpc_handler_test_server.h (hunk 5 above)
Comment (cschuet) [reply cartographer_grpc/framework/testing/rpc_handler_test_server.h]: DOne.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**